kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/metrics/TransactionReportSuite.scala

```diff
@@ -0,0 +1,378 @@
+          )),
+        path,
+        expectException = false,
+        expectedSnapshotVersion = 0,
+        expectedNumRemoveFiles = 1,
+        expectedNumTotalActions = 2, // commitInfo + removeFile
+        expectedCommitVersion = Some(1)
+      )
+    }
+  }
+
+  test("TransactionReport: retry with a concurrent append") {
+    withTempDir { tempDir =>
+      val path = tempDir.getCanonicalPath
+      // Set up delta table with version 0
+      spark.range(10).write.format("delta").mode("append").save(path)
+
+      checkTransactionReport(
+        generateCommitActions = (trans, engine) => {
+          spark.range(10).write.format("delta").mode("append").save(path)
+          generateAppendActions(fileStatusIter1)(trans, engine)
+        },
+        path,
+        expectException = false,
+        expectedSnapshotVersion = 0,
+        expectedNumAddFiles = 1,
+        expectedNumTotalActions = 2, // commitInfo + removeFile
+        expectedCommitVersion = Some(2),
+        expectedNumAttempts = 2
+      )
+    }
+  }
+
+  test("TransactionReport: fail due to conflicting write") {
+    withTempDir { tempDir =>
+      val path = tempDir.getCanonicalPath
+      // Set up delta table with version 0
+      spark.range(10).write.format("delta").mode("append").save(path)
+
+      checkTransactionReport(
+        generateCommitActions = (trans, engine) => {
+          spark.sql("ALTER TABLE delta.`" + path + "` ADD COLUMN newCol INT")
+          generateAppendActions(fileStatusIter1)(trans, engine)
+        },
+        path,
+        expectException = true,
+        expectedSnapshotVersion = 0
+      )
+    }
+  }
+
+  test("TransactionReport: fail due to too many tries") {
+    withTempDir { tempDir =>
+      val path = tempDir.getCanonicalPath
+      // Set up delta table with version 0
+      spark.range(10).write.format("delta").mode("append").save(path)
+
+      // This writes a concurrent append everytime the iterable is asked for an iterator. This means
+      // there should be a conflicting transaction committed everytime Kernel tries to commit
+      def actionsIterableWithConcurrentAppend(
+        trans: Transaction, engine: Engine): CloseableIterable[Row] = {
+        val transState = trans.getTransactionState(engine)
+        val writeContext = Transaction.getWriteContext(engine, transState, Collections.emptyMap())
+
+        new CloseableIterable[Row] {
+
+          override def iterator(): CloseableIterator[Row] = {
+            spark.range(10).write.format("delta").mode("append").save(path)
+            Transaction.generateAppendActions(engine, transState, fileStatusIter1, writeContext)
+          }
+
+          override def close(): Unit = ()
+        }
+      }
+
+      checkTransactionReport(
+        generateCommitActions = actionsIterableWithConcurrentAppend,
+        path,
+        expectException = true,
+        expectedSnapshotVersion = 0,
+        expectedNumAttempts = 200
```

Comment:
would be great if you could rebase this PR off of https://github.com/delta-io/delta/pull/4054 (once it is merged)

so that we don't retry 200 times but maybe more like 5. 200 is probably very slow.

Reply:
Lol I also was very confused when I wrote this test because I did not expect it to be 200!

Reply:
Could you work with @nicklan to get #4054 merged? Perhaps take over his PR if he is busy? After that, will approve this PR :)

Reply:
Merged & updated